Implement half-infinite projector algorithm

pbrehmer opened the pull request:
This PR will implement the full-infinite projector algorithm which SVDs the full 4x4 CTMRG environment, as suggested in #93.

To do this we will need to break up the `expand`/`projectors`/`renormalize` pattern we currently use to implement the CTMRG routine.

pbrehmer commented:
This should almost be finished. CTMRG is now refactored with both `HalfInfiniteProjector` and `FullInfiniteProjector` running, at least in the forward pass. Somehow `FullInfiniteProjector` does not yet differentiate but I'll look at that on Monday. I have also added the new projector algorithm to the test suite.

@lkdvos Perhaps you could take a look at the refactoring whenever you have time? Thanks :-)

lkdvos commented on src/algorithms/ctmrg/sequential.jl:
```diff
@@ -0,0 +1,109 @@
+    coordinates = eachcoordinate(envs)[:, col]
+    projectors = dtmap(coordinates) do (r, c)
+        proj, info = sequential_projectors((WEST, r, c), state, envs, alg)
+        ϵ = max(ϵ, info.err / norm(info.S))
```
Just FYI, I noticed this before too, but this really shouldn't be working: we are updating epsilon in parallel, which might lead to race conditions etc. The solution is to either use an atomic variable, or just also return this value.

pbrehmer replied:
Whoopsie, don't know why I missed that before. I quickly resolved that by also writing the errors into an pre-allocated Zygote Buffer.

lkdvos commented on src/algorithms/ctmrg/simultaneous.jl:
```diff
@@ -0,0 +1,108 @@
+        U[coordinate...] = info.U
+        S[coordinate...] = info.S
+        V[coordinate...] = info.V
+        ϵ = max(ϵ, info.err / norm(info.S))
```
same comment about race condition

lkdvos commented on src/algorithms/ctmrg/ctmrg.jl:
```diff
@@ -42,57 +44,77 @@ function svd_algorithm(alg::ProjectorAlg, (dir, r, c))
-Accessors.constructorof(::Type{CTMRG{S}}) where {S} = CTMRG{S}
+Perform a single CTMRG iteration in which all directions are being grown and renormalized.
+"""
+function ctmrg_iteration(state, env, alg::CTMRG)
+    if alg.flavor == :simultaneous
+        return simultaneous_ctmrg_iter(state, env, alg)
+    elseif alg.flavor == :sequential
+        return sequential_ctmrg_iter(state, env, alg)
+    end
```
just noticed that you put the flavor in a field, hence no more specializing on the alg struct. I'm not entirely sure if this really is more elegant, now that I see it

pbrehmer replied:
Now it dispatches on `SimultaneousCTMRG` or `SequentialCTMRG`

lkdvos reviewed (commented):
Overall, looks good!

I would say that I am a bit torn on using a flavor over a separate struct now, given that we are separating the algorithms in their own files and this results in having to come up with new names for functions that are doing the same thing, albeit via a different implementation. Maybe just something like this:
```julia
abstract type CTMRGAlgorithm end
struct SequentialCTMRG
  # fields....
end
struct SimultaneousCTMRG
end
```
could also work, although that is of course the same as the type aliases we had before. I like this slightly better since it doesn't require symbols, which are hard to constrain/remember the exact form of.

The projector code looks good, I think I would separate it into its own file, since it's now a substantial amount of code.

The remark about the race conditions is probably not super urgent, just wanted to at least mention it somewhere because I had noticed it and don't want to forget.

pbrehmer commented on src/algorithms/ctmrg/ctmrg.jl:
```diff
@@ -8,31 +8,33 @@ have different spaces, this truncation style is different from `TruncationSpace`
 
-function svd_algorithm(alg::ProjectorAlg, (dir, r, c))
+# TODO: do AbstractProjectorAlg type instead? -> would make it easier for users to implement custom projector alg
+const ProjectorAlgs = Union{HalfInfiniteProjector,FullInfiniteProjector}
```
Yes, after implementing it that way I also very much prefer having `ProjectorAlgorithm` as an abstract type, so that's what I did now.

pbrehmer commented on src/algorithms/contractions/ctmrg_contractions.jl:
```diff
@@ -210,13 +210,54 @@ function right_projector(E_1, C, E_2, U, isqS, ket::PEPSTensor, bra::PEPSTensor=
-                             ket_1::P, ket_2::P, bra_1::P=ket_1, bra_2::P=ket_2) where {P<:PEPSTensor}
-    halfinfinite_environment(x, C_1, C_2, E_1, E_2, E_3, E_4,
-                             ket_1::P, ket_2::P, bra_1::P=ket_1, bra_2::P=ket_2) where {P<:PEPSTensor}
+    left_and_right_projector(U, S, V, Q::AbstractTensorMap{E,3,3}, Q_next::AbstractTensorMap{E,3,3}
```
Completely agree, changed that to `contract_projectors`!

pbrehmer commented:
So most things should work now. Again, the truncation space coordinates for the projectors really confused the hell out of me but `FixedSpaceTruncation` now works properly again.

Somehow `FullInfiniteProjector` still doesn't work for the `:fixed` mode differentiation (the others seem to work), so I'll have to see tomorrow where the problem lies.

pbrehmer commented:
It should now differentiate also in the `:fixed` mode. It seems that the full-infinite projectors are just a bit unstable in general when it comes to differentiation (I think this is know?), so this leaves some room for experimentation. But from the implementation side this should be good to go, I hope.

I also updated the gauge-fixing test such that it hopefully is more stable now.

lkdvos reviewed (approved):
This looks good to me!

One remark might be that maybe we try to merge #97 first, since I don't want to make Yue have to rewrite his code (again). Could you have a look at that PR and tell us what you think?

pbrehmer commented:
Yes, I'll take a look and then we can merge #97 first!
